trin-core/src/utils/hash_delay_queue.rs

```diff
@@ -0,0 +1,395 @@
+            self.expirations.reset(delay_key, timeout);
+            true
+        } else {
+            false
+        }
+    }
+
+    /// Returns `true` if the queue contains an entry for `key`, `false` otherwise.
+    pub fn contains_key(&self, key: &K) -> bool {
+        self.entries.contains_key(key)
+    }
+
+    /// Returns the length of the queue.
+    #[allow(dead_code)]
+    pub fn len(&self) -> usize {
+        self.entries.len()
+    }
+
+    /// Returns whether the queue is empty.
+    #[allow(dead_code)]
+    pub fn is_empty(&self) -> bool {
+        self.entries.len() == 0
+    }
+
+    /// Removes the entry associated with the `key`. Returns `true` if there was an entry
+    /// associated with `key`, `false` otherwise.
+    pub fn remove(&mut self, key: &K) -> bool {
+        if let Some(delay_key) = self.entries.remove(key) {
+            self.expirations.remove(&delay_key);
+            true
+        } else {
+            false
+        }
+    }
+
+    /// Clears the queue, removing all entries.
+    #[allow(dead_code)]
+    pub fn clear(&mut self) {
+        self.entries.clear();
+        self.expirations.clear();
+    }
+}
+
+impl<K> Stream for HashDelayQueue<K>
+where
+    K: std::cmp::Eq + std::hash::Hash + std::clone::Clone + Unpin,
+{
+    type Item = Result<K, String>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        match self.expirations.poll_expired(cx) {
+            Poll::Ready(Some(Ok(key))) => match self.entries.remove(key.get_ref()) {
+                Some(_delay_key) => Poll::Ready(Some(Ok(key.into_inner()))),
+                None => Poll::Ready(Some(Err("Value no longer exists in expirations".into()))),
+            },
+            Poll::Ready(Some(Err(e))) => {
+                Poll::Ready(Some(Err(format!("delay queue error: {:?}", e))))
+            }
+            Poll::Ready(None) => Poll::Ready(None),
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    use std::collections::HashSet;
+
+    use tokio::time::{self, sleep, Duration};
+    use tokio_test::{assert_ok, assert_pending, assert_ready, task};
+
+    macro_rules! poll {
+        ($queue:ident) => {
+            $queue.enter(|cx, queue| queue.poll_next(cx))
+        };
+    }
+
+    macro_rules! assert_ready_ok_some {
+        ($e:expr) => {{
+            assert_ok!(match assert_ready!($e) {
+                Some(v) => v,
+                None => panic!("None"),
+            })
+        }};
+    }
+
+    const DELAY: Duration = Duration::from_millis(5);
+
+    #[tokio::test]
+    async fn insert_zero_delay() {
+        time::pause();
+
+        let mut queue = task::spawn(HashDelayQueue::<String>::new(DELAY));
+        let key = "foo".to_string();
+
+        queue.insert_with_timeout(key.clone(), Duration::ZERO);
+
+        sleep(Duration::from_millis(1)).await;
+
+        let entry = assert_ready_ok_some!(poll!(queue));
```

Comment:
It feels a little strange to me that an `assert_*` statement returns some kind of value... but I see that's also the behavior of some of these tokio `assert_*`s so it's consistent